IdentityCore/src/webview/embeddedWebview/MSIDOAuth2EmbeddedWebviewController.m

```diff
@@ -26,21 +26,375 @@
+#import "MSIDPKeyAuthHandler.h"
+#else
+#define DEFAULT_WINDOW_WIDTH 420
+#define DEFAULT_WINDOW_HEIGHT 650
+#endif
 
 @implementation MSIDOAuth2EmbeddedWebviewController
+{
+    NSURL *_startUrl;
+    NSURL *_endUrl;
+    void (^_completionHandler)(MSIDWebOAuth2Response *response, NSError *error);
+    
+    NSLock *_completionLock;
+    NSTimer *_spinnerTimer; // Used for managing the activity spinner
+}
 
+- (id)initWithStartUrl:(NSURL *)startUrl
+                endURL:(NSURL *)endUrl
+               webview:(WKWebView *)webview
+               context:(id<MSIDRequestContext>)context
+{
+    self = [super initWithContext:context];
+    
+    if (self)
+    {
+        self.webView = webview;
+        _startUrl = startUrl;
+        _endUrl = endUrl;
+        
+        _completionLock = [[NSLock alloc] init];
+        self.complete = NO;
+    }
+    
+    return self;
+}
+
+-(void)dealloc
+{
+    [self.webView setNavigationDelegate:nil];
+    self.webView = nil;
+}
+
+- (BOOL)startWithCompletionHandler:(MSIDWebUICompletionHandler)completionHandler
+{
+    // If we're not on the main thread when trying to kick up the UI then
+    // dispatch over to the main thread.
+    if (![NSThread isMainThread])
+    {
+        dispatch_async(dispatch_get_main_queue(), ^{
+            [self startWithCompletionHandler:completionHandler];
+        });
+        return YES;
+    }
+    
+    // Save the completion block
+    _completionHandler = [completionHandler copy];
+    
+    NSError *error = nil;
+    [self loadView:&error];
+    if (error)
+    {
+        [self endWebAuthenticationWithError:error orURL:nil];
+        return YES;
+    }
+    
+    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] initWithURL:_startUrl];
 #if TARGET_OS_IPHONE
-@synthesize  parentViewController;
+    // Currently Apple has a bug in iOS about WKWebview handling NSURLAuthenticationMethodClientCertificate.
+    // It swallows the challenge response rather than sending it to server.
+    // Therefore we work around the bug by using PKeyAuth for WPJ challenge in iOS
+    [request addValue:kMSIDPKeyAuthHeaderVersion forHTTPHeaderField:kMSIDPKeyAuthHeader];
 #endif
+    [self startRequest:request];
+    return YES;
+}
 
-- (void)startRequestWithCompletionHandler:(MSIDWebUICompletionHandler)completionHandler
+- (void)cancel
 {
+    MSID_LOG_INFO(self.context, @"Cancel Web Auth...");
     
+    // Dispatch the completion block
+    NSError *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorUserCancel, @"The user/application has cancelled the authorization.", nil, nil, nil, self.context.correlationId, nil);
+    [self endWebAuthenticationWithError:error orURL:nil];
 }
 
-- (void)cancel
+- (BOOL)loadView:(NSError **)error
+{
+    BOOL result = YES;
+    
+    if (!self.webView)
+    {
+        // create and load the view if not provided
+        result = [self createAndLoadView:error];
+    }
+    
+    self.webView.navigationDelegate = self;
```

Comment:
Just curious, do we need to remember the previous delegate?

Reply:
why? :)